src/Server/HTTPHandler.cpp

```diff
@@ -297,15 +388,65 @@ void HTTPHandler::processQuery(
     client_info.http_referer = request.get("Referer", "");
     client_info.forwarded_for = request.get("X-Forwarded-For", "");
 
-    /// This will also set client_info.current_user and current_address
-    context.setUser(user, password, request.clientAddress());
+    try
+    {
+        context.setUser(*request_credentials, request.clientAddress());
+    }
+    catch (const Authentication::Require<BasicCredentials> & required_credentials)
+    {
+        request_credentials = std::make_unique<BasicCredentials>();
+
+        if (required_credentials.getRealm().empty())
+            response.set("WWW-Authenticate", "Basic");
+        else
+            response.set("WWW-Authenticate", "Basic realm=\"" + required_credentials.getRealm() + "\"");
+
+        response.setStatusAndReason(HTTPResponse::HTTP_UNAUTHORIZED);
+        response.send();
+        return false;
+    }
+    catch (const Authentication::Require<GSSAcceptorContext> & required_credentials)
+    {
+        request_credentials = request_context->makeGSSAcceptorContext();
+
+        if (required_credentials.getRealm().empty())
+            response.set("WWW-Authenticate", "Negotiate");
+        else
+            response.set("WWW-Authenticate", "Negotiate realm=\"" + required_credentials.getRealm() + "\"");
+
+        response.setStatusAndReason(HTTPResponse::HTTP_UNAUTHORIZED);
+        response.send();
+        return false;
+    }
+
+    request_credentials.reset();
+
     if (!quota_key.empty())
         context.setQuotaKey(quota_key);
 
     /// Query sent through HTTP interface is initial.
     client_info.initial_user = client_info.current_user;
     client_info.initial_address = client_info.current_address;
 
+    return true;
+}
+
+
+void HTTPHandler::processQuery(
+    Context & context,
+    HTTPServerRequest & request,
+    HTMLForm & params,
+    HTTPServerResponse & response,
+    Output & used_output,
+    std::optional<CurrentThread::QueryScope> & query_scope)
+{
+    using namespace Poco::Net;
+
+    LOG_TRACE(log, "Request URI: {}", request.getURI());
+
+    if (!authenticateUser(context, request, params, response))
+        return; // '401 Unauthorized' response with 'Negotiate' has been sent at this point.
```

Comment:
If I understood correctly, `processQuery()` will be called two times for the same query. Is it right?

Reply:
If the HTTPHandler is not recreated - yes. Otherwise, if recreated, the second HTTP request will reach the second instance of HTTPHandler and will be made "correct" by the user, and will pass the auth checks, so only one call will take place for that last HTTPHandler instance.